### PR TITLE
Almost done port the dom transoform of babel-plugin-jsx-dom-expressions

### PR DESCRIPTION
**Pull request by huudev**
- follow babel-plugin-jsx-dom-expressions version v0.36.10.
- pass 10/11 test dom of babel-plugin-jsx-dom-expressions  except text-interpolation (because swc Evaluator is not complete, 
but it's fine to see the expression as dynamic).

### CODE REVIEW

**Comment by modderme123** [comment]
I just noticed a tiny bug in the ThisVisitor where `this` gets transformed in unexpected places, such as a store, which can cause issues (although amazingly that was the only issue with the solid playground, this is excellent work!).

```js
function createStore() {}

function Counter() {
  const [store, setStore] = createStore({
    a: 1,
    get b() {
      return this.a;
    },
  });

  return <button></button>;
}
```

I'm going to try and fix that this weekend but if there's any chance you have some idea that would be super helpful. Also, if you happen to be on our Discord (https://discord.gg/solidjs), I'd love to chat more there

File: src/shared/transform.rs [hunk 1]
```diff
@@ -1,70 +1,330 @@
...
+        if let Expr::This(_) = n {
+            if self.this_id.is_none() {
+                self.this_id = Some(self.this_block_visitor.generate_uid_identifier("self$"));
+            }
+            *n = Expr::Ident(self.this_id.clone().unwrap());
+        } else {
+            n.visit_mut_children_with(self);
+        }
+    }
+}
```
Comment (modderme123) [inline src/shared/transform.rs:152]:
```suggestion
pub struct ThisBlockVisitor {
    this_id: Option<Ident>,
    has_jsx: bool,
}

impl ThisBlockVisitor {
    pub fn new() -> Self {
        Self {
            this_id: None,
            has_jsx: false,
        }
    }
}

impl VisitMut for ThisBlockVisitor {
    fn visit_mut_block_stmt(&mut self, block: &mut BlockStmt) {
        let mut id = self.this_id.clone();
        self.has_jsx = false;
        block.visit_mut_children_with(self);
        std::mem::swap(&mut self.this_id, &mut id);
        if let Some(id) = id {
            block.stmts.insert(
                0,
                Stmt::Decl(Decl::Var(Box::new(VarDecl {
                    span: DUMMY_SP,
                    kind: VarDeclKind::Const,
                    declare: false,
                    decls: vec![VarDeclarator {
                        span: DUMMY_SP,
                        name: Pat::Ident(id.into()),
                        init: Some(Box::new(Expr::This(ThisExpr { span: DUMMY_SP }))),
                        definite: false,
                    }],
                }))),
            )
        }
    }
    fn visit_mut_jsx_element(&mut self, el: &mut JSXElement) {
        self.has_jsx = true;
        el.visit_mut_children_with(self);
    }
    fn visit_mut_jsx_fragment(&mut self, el: &mut JSXFragment) {
        self.has_jsx = true;
        el.visit_mut_children_with(self);
    }
    fn visit_mut_expr(&mut self, n: &mut Expr) {
        if let Expr::This(_) = n {
            if !self.has_jsx {
                return;
            }
            if self.this_id.is_none() {
                self.this_id = Some(private_ident!("_self$"));
            }
            *n = Expr::Ident(self.this_id.clone().unwrap());
        } else {
            n.visit_mut_children_with(self);
        }
    }
}
```

**Comment by modderme123** [comment]
Update: I believe I've fixed the `this` issue (see suggestion above), and ran into a runtime panic with running

```js
const asdf = (
  <div>
    <div use:something>hi</div>
  </div>
);
```

File: src/shared/utils.rs [hunk 2]
```diff
@@ -211,3 +631,174 @@ pub fn wrapped_by_text(list: &[ImmutableChildTemplateInstantiation], start_index
...
+
+pub fn jsx_text_to_str(t: &Atom) -> JsWord {
+    let mut buf = String::new();
+    let replaced = t.replace('\t', " ");
+
+    for (is_last, (i, line)) in replaced.lines().enumerate().identify_last() {
+        if line.is_empty() {
+            continue;
+        }
+        let line = Cow::from(line);
```
Comment (modderme123) [inline src/shared/utils.rs:784]: It seems like you can safely remove Cow here

**Comment by modderme123** [comment]
Ok the remaining two bugs seem to be an issue handling a 690kb file with a wasm plugin (it works in the test harness but not with @swc/core) and a slight difference where we compile text nodes to a template and the babel plugin doesn't:

```ts
const x = <>{'bla'}</>
```
should compile to
```ts
const x = 'bla';
```

**Comment by modderme123** [comment]
Oops, https://fork.dev has a bug that caused this to be prematurely merged. In any case, thanks for the awesome PR! I'll keep working on the two issues above but if you have any ideas that would be great

File: src/shared/transform.rs (hunk 1 above)
Comment (huudev) [reply src/shared/transform.rs:152]: Thank you for fixing this issue. Nice code.

**Pull request merged**